### ISSUE TITLE
Invoke-SSHCommand slow to many hosts over slow link

### ISSUE THREAD

**Issue by FireInWinter** [darkoperator/Posh-SSH#55]
Invoke-SSHCommand is slow if you are running against a number of systems over a slow link.  Looking at the Invoke-SSHCommand code, it looks like it should be possible to do an BeginExecute on each connection, assign the results to an array, then iterate through them in the "end" section of the script.  This way they would execute in parallel, but you could get the results from them after the fact.  This should make some scripts much faster when ran against multiple systems.

**Comment by darkoperator** [comment]
That would mean me following state per connectiononly working with x number of connections at once, finish those, grab the next ones and probably working with runsspaces. A change that big with my current workload would not even be possible for months for me to experiment. But I do accept pull requests ;)

**Comment by FireInWinter** [comment]
Shouldn't have to go to the extreme of using runspaces.  You already are using async methods, it just has to go into an array.  But I'll try to get things setup so I can put in a pull request.  Haven't done anything with git yet, so it will take me a bit to figure it out.

**Comment by darkoperator** [comment]
So far my experience is that powershell sucks at asynchronous tasks and they have to be handled by jobs, runsspaces or as a cmdlet in C#. Also if running against multiple sessions it should have a control to run against x amount of sessions at once like other cmdlet a that have a limit of 32 parallel tasks at a time.

**Comment by FireInWinter** [comment]
Well.  I had some time so I put together a quick version that does it async.  It doesn't have limits on parallel tasks, but it is using remote threads instead of local.  And the main cost with async of establishing the connection is already done.  I use the variable $AsyncProcessing to keep track of the different connections and their status.   In my case, it took getting results from over 200 servers from 25 seconds to 9 seconds.  It should be better, but it isn't really multi-threaded.  So it has to start the commands sequentially and then end them again sequentially. 

I'm actually not sure how bad it is not to have a threshold on it.  Long term it could be added, but short term it is more useful to have it faster, but not a perfect design.  The biggest reason to limit parallelism, is that resources are consumed locally to set up all the connections.  In this case, the connections are already established, so the cost doesn't really come into play.

```powershell
function Invoke-SSHCommand
{
    [CmdletBinding(DefaultParameterSetName='Index')]
    param(

        [Parameter(Mandatory=$true,
                   Position=1)]
        [string]$Command,
        
        [Parameter(Mandatory=$true,
                   ParameterSetName = 'Session',
                   ValueFromPipeline=$true,
                   Position=0)]
        [Alias('Name')]
        [SSH.SSHSession[]]
        $SSHSession,

        [Parameter(Mandatory=$true,
                   ParameterSetName = 'Index',
                   Position=0)]
        [Alias('Index')]
        [int32[]]
        $SessionId = $null,

        # Ensures a connection is made by reconnecting before command.
        [Parameter(Mandatory=$false)]
        [switch]
        $EnsureConnection,

        [Parameter(Mandatory=$false,
                   Position=2)]
        [int]
        $TimeOut = 60

    )

    Begin
    {
        $ToProcess = @()
        switch($PSCmdlet.ParameterSetName)
        {
            'Session'
            {
                $ToProcess = $SSHSession
            }

            'Index'
            {
                foreach($session in $Global:SshSessions)
                {
                    if ($SessionId -contains $session.SessionId)
                    {
                        $ToProcess += $session
                    }
                }
            }
        }

        # array to keep track of all the running jobs.
        $AsyncProcessing = @()
    }
    Process
    {
        foreach($Connection in $ToProcess)
        {
            if ($Connection.Session.IsConnected)
            {
                if ($EnsureConnection)
                {
                    try
                    {
                        $Connection.Session.Connect()
                    }
                    catch
                    {                            
                        if ( $_.Exception.InnerException.Message -ne 'The client is already connected.' )
                        { Write-Error -Exception $_.Exception }
                    }
                }
                #$cmd = $Connection.session.CreateCommand($Command)
            }
            else
            {
                try
                {
                    $Connection.Session.Connect()
                }
                catch
                {
                    Write-Error "Unable to connect session to $($Connection.Host) :: $_"
                }
                
            }

            if ($Connection.Session.IsConnected)
            {
                $cmd = $Connection.session.CreateCommand($Command)
                $cmd.CommandTimeout = [timespan]::FromMilliseconds(50) #New-TimeSpan -Seconds $TimeOut

                # start asynchronious execution of the command.
                $Duration = [System.Diagnostics.Stopwatch]::StartNew()
                $Async = $cmd.BeginExecute()
                
                $AsyncProcessing  += [pscustomobject]@{
                    cmd = $cmd
                    Async = $Async
                    Connection = $Connection
                    Duration = $Duration
                    Processed = $false
                }
            }
        }
    }
    End
    {
        $Done = $false
        while ( -not $Done )
        {
            $AsyncProcessing |
            % {
                # Check if it completed or is past the timeout setting.
                if ( $_.Async.IsCompleted -or $_.Duration.Elapsed.TotalSeconds -gt $TimeOut )
                {
                    $Output = $_.cmd.EndExecute($_.Async)
                    
                    # Generate custom object to return to pipeline and client
                    [pscustomobject]@{
                        Output = $Output
                        ExitStatus = $_.cmd.ExitStatus
                        Error = $_.cmd.Error
                        Host = $_.Connection.Host
                        Duration = $_.Duration.Elapsed
                    } | 
                    % { 
                        $_.pstypenames.insert(0,'Renci.SshNet.SshCommand'); 

                        #Return object to pipeline
                        $_
                    }

                    # Set this object as having been processed.
                    $_.Processed = $true
                }
            }

            # Check if the list is done or sleep for 50 ms.
            $AsyncProcessing = @($AsyncProcessing | ? { -not $_.Processed } )
            if ( $AsyncProcessing.Count -eq 0 )
            { $Done = $true }
            else { Start-Sleep -Milliseconds 50 }
        }
    }
}
```
